### PR TITLE
Deduplicate names in gloss; gloss towns

### PR DESCRIPTION
**Pull request by 1ec5**
Replaced the general-purpose function for generating string replacement expressions with a function for generating an expression that specifically scans a string for semicolon delimiters, escaped semicolons, and space padding after a semicolon. Compared to #666, this change decreases the overall minified style JSON size[^size] (a rough proxy for style complexity) from 954,393&nbsp;characters down to just 932,343&nbsp;characters.

With this optimization in place, a city’s local-language name gloss now omits any name that matches the city’s name in the user’s preferred language to avoid repetition. This brings the size back up to 943,759&nbsp;characters. Unfortunately, every multilingual `place=city` is currently tagged with a `name` that contains delimiters other than a semicolon, making this enhancement rather pointless. To make this deduplication worthwhile, this PR also enables a local-name gloss on `town`s too.

Previously, I mentioned in https://github.com/ZeLonewolf/openstreetmap-americana/issues/471#issue-1292150160 that American maps and atlases only gloss major city labels, but we’ve already been glossing every `city` label. Spot-checking shows that label density isn’t suffering too much from the more verbose town labels, and in any case we support zooming in. The one downside to enabling glosses on town labels is that the town label layer has to repeat the massive expression that we’ve been using for cities: we’re back up to 952,745&nbsp;characters&nbsp;– still less than we started out with, but only slightly.

[Kaser and New Square](https://zelonewolf.github.io/openstreetmap-americana/#map=12/41.13986/-74.0432&language=en)

<table>
<thead>
<tr>
<th>Change</th><th>Style JSON size (characters)</th>
</tr>
</thead>
<tbody>
<tr>
<td>replacer</td><td align="right">954,393</td>
</tr>
<tr>
<td>− replacer + scanner</td><td align="right">−22,050</td>
</tr>
<tr>
<td>+ deduplication</td><td align="right">+11,416</td>
</tr>
<tr>
<td>+ town glosses</td><td align="right">+8,986</td>
</tr>
</tbody>
<tfoot>
<tr>
<th scope="row">Net change</th><td align="right">−1,648</td>
</tfoot>
</table>

/ref https://github.com/ZeLonewolf/openstreetmap-americana/pull/592#discussion_r1035284310 https://github.com/ZeLonewolf/openstreetmap-americana/pull/666#issuecomment-1372443684

[^size]: `JSON.stringify(map.getStyle()).length`

### CODE REVIEW

File: src/constants/label.js [hunk 1]
```diff
@@ -130,42 +129,118 @@ export function localizeLayers(layers, locales) {
...
   if (numReplacements <= 0) {
     return asIs;
   }
 
-  let needleStart = ["index-of", needle, haystack, haystackStart];
-  let needleLength =
-    typeof needle === "object" ? ["length", needle] : needle.length;
-  let needleEnd = ["+", needleStart, needleLength];
+  let iteration = numReplacements;
+  let rawSeparator = ";";
```
Comment (1ec5) [inline src/constants/label.js:151]: I spilled a lot of ink in [this forum discussion](https://community.openstreetmap.org/t/multiple-delimited-names-in-the-name-tag/6803) about the need for a reliable delimiter&nbsp;– the semicolon. But what if we pretend that one of the popular ad-hoc delimiters is unambiguous enough to split on instead?

Sad reality | Simulation
----|----
[Alicante before](https://zelonewolf.github.io/openstreetmap-americana/#map=8.65/38.3721/-0.4708&language=en) | [Alicante after](https://zelonewolf.github.io/openstreetmap-americana/#map=8.65/38.3721/-0.4708&language=en)
[Brussels before](https://zelonewolf.github.io/openstreetmap-americana/#map=12/50.86018/4.34901&language=en) | [Brussels after](https://zelonewolf.github.io/openstreetmap-americana/#map=12/50.86018/4.34901&language=en)
[Casablanca before](https://zelonewolf.github.io/openstreetmap-americana/#map=8/33.578/-7.378&language=en) | [Casablanca after](https://zelonewolf.github.io/openstreetmap-americana/#map=8/33.578/-7.378&language=en)
[Hong Kong before](https://zelonewolf.github.io/openstreetmap-americana/#map=10.5/22.3253/114.217&language=en) | [Hong Kong after](https://zelonewolf.github.io/openstreetmap-americana/#map=10.5/22.3253/114.217&language=en)

But alas, we can’t just turn every hyphen or slash or space into a value delimiter:

Sad reality | Sadder simulation
----|----
[Tizi Ouzou and Bordj Bou Arreridjj before](https://zelonewolf.github.io/openstreetmap-americana/#map=7/36.284/4.355&language=en) | [Tizi Ouzou and Bordj Bou Arreridjj after](https://zelonewolf.github.io/openstreetmap-americana/#map=7/36.284/4.355&language=en)
[Reynoldsburg - New Albany Road before](https://zelonewolf.github.io/openstreetmap-americana/#map=14/40.06114/-82.8109/45&language=en) | [Reynoldsburg - New Albany Road after](https://zelonewolf.github.io/openstreetmap-americana/#map=14/40.06114/-82.8109/45&language=en)

**Comment by ZeLonewolf** [comment]
Great work! 

For places with multiple glosses, would it be better if each label has its own parentheses on a separate row?  Just thinking that the open parenthesis on one line and the close parenthesis on the last line looks a touch uneven.

**Comment by 1ec5** [comment]
> For places with multiple glosses, would it be better if each label has its own parentheses on a separate row? Just thinking that the open parenthesis on one line and the close parenthesis on the last line looks a touch uneven.

This is probably only hypothetical for now, since so few places with long names or many local languages are using the semicolon in their `name`s. But you’re right that there’s an imbalance. There’s not much we can do about that without GL JS support for [hanging punctuation](https://en.wikipedia.org/wiki/Hanging_punctuation) or [optical margin alignment](https://en.wikipedia.org/wiki/Optical_margin_alignment). A separate pair of parentheses on each line would look busy (really) (no, really really), and we’d still have some long names get word-wrapped onto multiple lines without any opportunities to introduce parentheses.

The only practical way to resolve this imbalance would be to nix the parentheses and rely on font attributes to distinguish the gloss from the main text. Some maps do rely on italics or a horizontal bar to set off the gloss, but I’m not sure it would work as well in this style, especially since we allow the label to appear above the icon.

File: src/constants/label.js [hunk 2]
```diff
@@ -385,7 +442,10 @@ export const localizedNameWithLocalGloss = [
       // bother rendering it.
       ["concat", ["slice", ["var", "localizedName"], 0, 1], " "],
       { "font-scale": 0.001 },
-      listValuesExpression(["get", "name"], " \u2022 "),
+      listValuesExpression(["get", "name"], " \u2022 ", [
```
Comment (ZeLonewolf) [inline src/constants/label.js:445]: Perhaps the bullet character could be extracted to a constant for better readability?

**Comment by ZeLonewolf** [comment]
When localized to Hebrew:
image

I assume this is a tagging error?

**Comment by 1ec5** [comment]
“<bdi>שיכון סקווירא</bdi>” is New Square’s name in Yiddish. It has a different name in Hebrew, “<bdi>ניו סקוור</bdi>”. #586 tracks the lack of a dedicated Yiddish name field in the vector tiles.

**Review by ZeLonewolf — APPROVED** [review]
This looks awesome, thanks!
